blockstore/splitstore/debug.go

```diff
@@ -0,0 +1,360 @@
+	}
+
+	d.writeMx.Lock()
+	defer d.writeMx.Unlock()
+
+	d.writeCnt += len(blks)
+
+	now := d.timestamp()
+	for _, blk := range blks {
+		_, err := fmt.Fprintf(d.writeLog, "%s %s%s\n", now, blk.Cid(), stack)
+		if err != nil {
+			log.Warnf("error writing write log: %s", err)
+			break
+		}
+	}
+}
+
+func (d *debugLog) LogDelete(cids []cid.Cid) {
+	if d == nil {
+		return
+	}
+
+	d.deleteMx.Lock()
+	defer d.deleteMx.Unlock()
+
+	d.deleteCnt += len(cids)
+
+	now := d.timestamp()
+	for _, c := range cids {
+		_, err := fmt.Fprintf(d.deleteLog, "%s %s\n", now, c)
+		if err != nil {
+			log.Warnf("error writing delete log: %s", err)
+			break
+		}
+	}
+}
+
+func (d *debugLog) Flush() {
+	if d == nil {
+		return
+	}
+
+	// rotate non-empty logs
+	d.rotateReadLog()
+	d.rotateWriteLog()
+	d.rotateDeleteLog()
+	d.rotateStackLog()
+}
+
+func (d *debugLog) rotateReadLog() {
+	d.readMx.Lock()
+	defer d.readMx.Unlock()
+
+	if d.readCnt == 0 {
+		return
+	}
+
+	err := d.rotate(d.readLog, d.readPath)
+	if err != nil {
+		log.Warnf("error rotating read log: %s", err)
+		return
+	}
+
+	d.readLog, err = os.OpenFile(d.readPath, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Warnf("error opening log file: %s", err)
+		return
+	}
+
+	d.readCnt = 0
+}
+
+func (d *debugLog) rotateWriteLog() {
+	d.writeMx.Lock()
+	defer d.writeMx.Unlock()
+
+	if d.writeCnt == 0 {
+		return
+	}
+
+	err := d.rotate(d.writeLog, d.writePath)
+	if err != nil {
+		log.Warnf("error rotating write log: %s", err)
+		return
+	}
+
+	d.writeLog, err = os.OpenFile(d.writePath, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Warnf("error opening write log file: %s", err)
+		return
+	}
+
+	d.writeCnt = 0
+}
+
+func (d *debugLog) rotateDeleteLog() {
+	d.deleteMx.Lock()
+	defer d.deleteMx.Unlock()
+
+	if d.deleteCnt == 0 {
+		return
+	}
+
+	err := d.rotate(d.deleteLog, d.deletePath)
+	if err != nil {
+		log.Warnf("error rotating delete log: %s", err)
+		return
+	}
+
+	d.deleteLog, err = os.OpenFile(d.deletePath, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Warnf("error opening delete log file: %s", err)
+		return
+	}
+
+	d.deleteCnt = 0
+}
+
+func (d *debugLog) rotateStackLog() {
+	d.stackMx.Lock()
+	defer d.stackMx.Unlock()
+
+	if d.stackCnt == 0 {
+		return
+	}
+
+	err := d.rotate(d.stackLog, d.stackPath)
+	if err != nil {
+		log.Warnf("error rotating stack log: %s", err)
+		return
+	}
+
+	d.stackLog, err = os.OpenFile(d.stackPath, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Warnf("error opening stack log file: %s", err)
+		return
+	}
+
+	d.stackCnt = 0
+}
+
+func (d *debugLog) rotate(f *os.File, path string) error {
+	err := f.Close()
+	if err != nil {
+		return xerrors.Errorf("error closing file: %w", err)
+	}
+
+	arxivPath := fmt.Sprintf("%s-%d", path, time.Now().Unix())
+	err = os.Rename(path, arxivPath)
+	if err != nil {
+		return xerrors.Errorf("error moving file: %w", err)
+	}
+
+	go func() {
+		cmd := exec.Command("gzip", arxivPath)
```

Comment:
https://golang.org/pkg/compress/gzip/?

Reply:
it is less code this way, but sure we can do it in process.

Reply:
You know, it's debug code. Up to you. I just cringe at external deps because they tend to lead to random bug reports.

Reply:
Yeah, it's immaterial. If a user doesn't have gzip in his system... what can I say? He'll have to live with some garbage during debugging.

Regardless, If I don't have anything better to do, I will turn it to internal code.